android/modules/calendar/src/java/ti/modules/titanium/calendar/RecurrenceRuleProxy.java

```diff
@@ -0,0 +1,376 @@
+	//endregion
+
+	// enum for matching Titanium frequency constants with Android recurrence rule words
+	public enum TiRecurrenceFrequencyType
+	{
+		DAILY(CalendarModule.RECURRENCEFREQUENCY_DAILY, "DAILY"),
+		WEEKLY(CalendarModule.RECURRENCEFREQUENCY_WEEKLY, "WEEKLY"),
+		MONTHLY(CalendarModule.RECURRENCEFREQUENCY_MONTHLY, "MONTHLY"),
+		YEARLY(CalendarModule.RECURRENCEFREQUENCY_YEARLY, "YEARLY");
+
+		private final int tiIntId;
+		private final String rfcStringId;
+
+		private TiRecurrenceFrequencyType(int tiIntId, String rfcStringId)
+		{
+			this.tiIntId = tiIntId;
+			this.rfcStringId = rfcStringId;
+		}
+
+		public int toTiIntId()
+		{
+			return this.tiIntId;
+		}
+
+		public String toRfcStringId()
+		{
+			return this.rfcStringId;
+		}
+
+		public static TiRecurrenceFrequencyType fromTiIntId(int value)
+		{
+			for (TiRecurrenceFrequencyType nextObject : TiRecurrenceFrequencyType.values()) {
+				if ((nextObject != null) && (nextObject.tiIntId == value)) {
+					return nextObject;
+				}
+			}
+			return null;
+		}
+
+		public static TiRecurrenceFrequencyType fromRfcStringId(String value)
+		{
+			for (TiRecurrenceFrequencyType nextObject : TiRecurrenceFrequencyType.values()) {
+				if ((nextObject != null) && (nextObject.rfcStringId == value)) {
+					return nextObject;
+				}
+			}
+			return null;
+		}
+	}
+
+	// Map matching days of the week constants from Titanium docs with their String counterparts in RRULE column.
+	private static final Map<String, Integer> weekdaysMap;
+	static
+	{
+		weekdaysMap = new HashMap<String, Integer>();
+		weekdaysMap.put("SU", 1);
+		weekdaysMap.put("MO", 2);
+		weekdaysMap.put("TU", 3);
+		weekdaysMap.put("WE", 4);
+		weekdaysMap.put("TH", 5);
+		weekdaysMap.put("FR", 6);
+		weekdaysMap.put("SA", 7);
+	}
+
+	// Constructor for creating a recurrence rule from a Kroll creation dictionary.
+	public RecurrenceRuleProxy(KrollDict creationDictionary) {
+		if (creationDictionary.containsKey(TiC.PROPERTY_CALENDAR_ID)) {
+			this.calendarID = TiConvert.toString(creationDictionary.get(TiC.PROPERTY_CALENDAR_ID));
+		}
+		if (creationDictionary.containsKey(TiC.PROPERTY_CALENDAR_DAYS_OF_THE_MONTH)) {
+			this.daysOfTheMonth = creationDictionary.getIntArray(TiC.PROPERTY_CALENDAR_DAYS_OF_THE_MONTH);
+		}
+		if (creationDictionary.containsKey(TiC.PROPERTY_CALENDAR_DAYS_OF_THE_WEEK)) {
+			this.daysOfTheWeek = creationDictionary.getKrollDictArray(TiC.PROPERTY_CALENDAR_DAYS_OF_THE_WEEK);
+		}
+		if (creationDictionary.containsKey(TiC.PROPERTY_CALENDAR_DAYS_OF_THE_YEAR)) {
+			this.daysOfTheYear = creationDictionary.getIntArray(TiC.PROPERTY_CALENDAR_DAYS_OF_THE_YEAR);
+		}
+		if (creationDictionary.containsKey(TiC.PROPERTY_END)) {
+			this.endDictionary = (KrollDict) creationDictionary.getKrollDict(TiC.PROPERTY_END);
+		}
+		if (creationDictionary.containsKey(TiC.PROPERTY_FREQUENCY)) {
+			this.frequency = TiRecurrenceFrequencyType.fromTiIntId(TiConvert.toInt(creationDictionary.get(TiC.PROPERTY_FREQUENCY)));
```

Comment:
The static `fromTiIntId()` method will return null if the given ID is invalid/unknown. We should either default to DAILY or throw an exception.

I've tested an invalid ID on iOS and it accepts the value as-is and puts the recurrence rule in an invalid state.

Perhaps throwing an exception would be better? It might be better to error out instead of assuming DAILY and creating unintended calendar events on the end-user's device. Note that throwing an exception is only valuable if thrown on the JavaScript runtime thread because we'll log a JavaScript stack trace.

Reply:
Not sure what would be the better way. I will leave at printing an Error message and not adding the wrong rule.